pkg/ingester/user_metrics_metadata_test.go

```diff
@@ -0,0 +1,113 @@
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "baz"}: false,
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "qux"}: false,
+			},
+			expectedMetadata: []*mimirpb.MetricMetadata{
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "foo"},
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "bar"},
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "baz"},
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "qux"},
+			},
+		},
+		"should fail when metadata per user limit reached": {
+			maxMetadataPerUser: 1,
+			metadataToExpectError: map[*mimirpb.MetricMetadata]bool{
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "foo"}: false,
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "bar"}: false,
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "baz"}: true,
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "qux"}: true,
+			},
+			expectedMetadata: []*mimirpb.MetricMetadata{
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "foo"},
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "bar"},
+			},
+		},
+		"should fail when metadata per metric limit reached": {
+			maxMetadataPerMetric: 1,
+			metadataToExpectError: map[*mimirpb.MetricMetadata]bool{
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "foo"}: false,
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "bar"}: true,
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "baz"}: false,
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "qux"}: true,
+			},
+			expectedMetadata: []*mimirpb.MetricMetadata{
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_1", Help: "foo"},
+				{Type: mimirpb.COUNTER, MetricFamilyName: "test_metric_2", Help: "baz"},
+			},
+		},
+	}
+
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			// Mock the ring
+			ring := &ringCountMock{}
+			ring.On("HealthyInstancesCount").Return(1)
+			ring.On("ZonesCount").Return(1)
+
+			// Mock limiter
+			limits, err := validation.NewOverrides(validation.Limits{
+				MaxGlobalMetricsWithMetadataPerUser: testData.maxMetadataPerUser,
+				MaxGlobalMetadataPerMetric:          testData.maxMetadataPerMetric,
+			}, nil)
+			require.NoError(t, err)
+			limiter := NewLimiter(limits, ring, 1, false)
+
+			// Mock metrics
+			metrics := newIngesterMetrics(
+				prometheus.NewPedanticRegistry(),
+				true,
+				func() *InstanceLimits { return defaultInstanceLimits },
+				nil,
+				nil,
+			)
+
+			mm := newMetadataMap(limiter, metrics, "test")
+
+			// Attempt to add all metadata
+			for meta, expectErr := range testData.metadataToExpectError {
+				err := mm.add(meta.MetricFamilyName, meta)
+
+				if expectErr {
+					require.Error(t, err)
+				}
```

Comment:
I would also check the other case. If we don't expect any error, we should assert on it.

```suggestion
				}  else {
					require.NoError(t, err)
				}
```